### ISSUE TITLE
Equivalent of Torch EmbeddingBag

### ISSUE THREAD

**Issue by Enumaris** [tensorflow/tensorflow#32675]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.14
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
Currently Tensorflow has only tf.nn.embedding_lookup (which, I believe in the backend uses a tf.gather operation). Torch has a torch.nn.EmbeddingBag class which gathers and combines embeddings in specified ways. It allows for a weighted summation of embeddings which they say is much more efficient than getting the embeddings, multiplying (broadcasting) a set of weights across them, and then reduce-summing them. 

I am trying to implement product-key memory layers (https://arxiv.org/pdf/1907.05242.pdf) in tensorflow but they are running very slowly for me - roughly 50% slower than an equivalent BERT-Base sized model without product-key memory layers. Looking at the debugging results, I find that the operations of tf.nn.embedding_lookup, the broadcasted multiplication between scores and values, and the tf.math.reduce_sum operation are taking up a lot of time. A more efficient implementation that performs these operations in one go should reduce the time it takes to run these layers. All product-key memory layers which have a weighted summation of some memory values should benefit from this optimization. 

For a concrete example pseudocode in tensorflow I have to do these operations:

query = tf.Tensor()
key = tf.Tensor()
values = tf.get_variable()
scores, indices = get_scores_and_indices(query,key)
attended_values = tf.nn.embedding_lookup(values,indices)
output = tf.math.reduce_sum(scores*attended_values)  

On torch, you could do
query = torch.Tensor()
key = torch.Tensor()
values = torch.nn.EmbeddingBag()
scores, indices = get_scores_and_indices(query,key)
output = values(indices, per_sample_weights=scores)

Torch claims that using an EmbeddingBag is much more efficient than using a simple Embedding+weighted sum operation. And this does indeed appear to be the case. 
 
**Will this change the current api? How?**
It will add new functionality in a tf.nn.embedding_bag_lookup() method

**Who will benefit with this feature?**
Anybody who wants to implement query-key memory layers based on values saved in embeddings. 

**Any Other info.**
The closest thing I could find to this was tf.nn.embedding_lookup_sparse, but that method requires the indices and scores to be sparse tensors and then just uses a tf.nn.embedding_lookup with some broadcast multiplication and reduce_sum in the backend anyhow.

**Comment by kobenaxie** [comment]
Same questions, currently my solution same with you ~
`
  scores, indices = get_scores_and_indices(query,key)  
`
`
  output = tf.nn.embedding_lookup(values, indices) # [bs, heads*knn, v_dim]
`
`
  scores = tf.expand_dims(scores, axis=-1) #[bs, heads*knn, 1]
`
`
  output = tf.reduce_sum(output * scors)   #[bs, v_dim]
`
any difference ? @Enumaris

**Comment by Enumaris** [comment]
@kobenaxie we are doing the same thing. I didn't explicitly write out the `scores = tf.expand_dims(scores, axis=-1)`  line of my code and I use a `reshape` instead of an `expand_dims` but otherwise we are doing identical operations.

**Comment by kobenaxie** [comment]
@Enumaris actually identical, sorry for not specifying the axis in reduce_sum
`output = tf.reduce_sum(output * scors, axis=1) #[bs, v_dim]`

**Comment by Enumaris** [comment]
@kobenaxie Right, we are performing identical operations. After some debugging, it appears that the `tf.math.reduce_sum(output*scores, axis=1)` line takes up all the time actually. These are quite large tensors that need to be broadcasted and multiplied together then summed. But it is beyond my scope of knowledge to figure out how to optimize these operations.

**Comment by gaceladri** [comment]
Could any of you reproduce the product-key memory layer in tf? I get here looking for the same, coming from the same place. Did you found a solution/implementation to this that works reasonably fast?

**Comment by Rocketknight1** [comment]
Hey everyone, I have a public reimplementation of the layer in TF at https://github.com/Rocketknight1/tf-pkm-layer . I've verified that output is identical to the PyTorch layer, and speed seems okay. My replacement for `EmbeddingBag` was

```
values = tf.gather(self.values, indices)
output = tf.einsum('ijk, ij -> ik', values, scores)  # (bs,v_dim)
``` 

The second line is equivalent to `tf.reduce_sum(values * tf.expand_dims(scores, -1))`. Performance seems reasonably good for me on GPU, but I haven't done extensive testing and integration into a bigger model yet.

**Comment by Rocketknight1** [comment]
Update: I tried benchmarking my implementation and when compiled with XLA, Tensorflow's performance for the PKM layers is only about 20% slower than PyTorch's, even though it doesn't have a native implementation of EmbeddingBag. Given that PKM layers are going to be a small fraction of the total runtime compared to other layers like self-attention, I'd guess full model performance should be almost identical between the two frameworks.

**Comment by gaceladri** [comment]
Did you try to reproduce the paper results? Did you gain some performance adding the pkm layer? Thanks in advance for the work!

**Comment by Rocketknight1** [comment]
@gaceladri I haven't tried retraining a full transformer model with it yet. In testing I realized that memory usage during training (but not inference) was very high, and the reason is that when you need a backward pass Tensorflow retains the very large tensor containing all the slices from the memory, before they are weighted and summed.

In PyTorch, this isn't a problem because all of that step happens in the native EmbeddingBag layer. I'm in the middle of writing a custom backward pass for my Tensorflow EmbeddingBag equivalent, so hopefully that fixes everything!

If I get it working, I'll push it to the repo and tag you here.

**Comment by Rocketknight1** [comment]
Update: I haven't been able to perfectly match the memory usage of Torch on this. The closest I got was using a tf.while_loop with a custom backward pass, to ensure that the whole pre-sum gathered tensor wasn't materialized all at the same time. Even with this method, though, we go OOM on batch sizes about half of the ones Torch can manage. I think doing better would require me diving into CUDA and adding a custom op, but I don't understand CUDA or the TF codebase nearly well enough to do that.

I'll tidy everything up and push the while loop implementation to my repo soon.

**Comment by gaceladri** [comment]
Thanks for the update. I directly changed my project to Pytorch...

**Comment by Rocketknight1** [comment]
Hey, I think I have a solution that has similar memory usage to Pytorch, at least up to about 500m network parameters. Beyond that point, the sheer size of the values and values gradient tensors gets really big, and PyTorch does a bit better for some reason.

The solution is very similar to my einsum solution above, but with some reshaping and use of map_fn with limited parallel steps and control_dependencies to keep memory usage down during the embeddingbag portion. I'm still working on performance - XLA ignores a lot of dependency control, so it blows out my memory again, but when I can use XLA performance is comparable to PyTorch.

**Comment by Rocketknight1** [comment]
I'll do some performance testing and then finally push everything soon.

**Comment by Rocketknight1** [comment]
Update: I was very wrong about that solution being a good one, but I've implemented a custom op in CUDA and am trying to push it into tf/addons right now.

**Comment by Rocketknight1** [comment]
See https://github.com/tensorflow/addons/issues/2201

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
